File: NonsenseView.m
//
//  NonsenseView.m
//  NonsenseSaver
//
//  Created by C.W. Betts on 9/14/10.
//  Copyright 2010 __MyCompanyName__. All rights reserved.
//

#import "NonsenseView.h"
#import "NonsenseSaverController.h"
#import "NonsenseObject.h"

@interface NonsenseView ()
@property (strong) NonsenseSaverController *nonsenseController;
@property (strong) NSMutableArray *nonsenses;
@property (strong) NSTimer *refreshRate;
@property BOOL settingsChanged;
@end

@implementation NonsenseView

- (id)initWithFrame:(NSRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
		self.nonDuration = 2.7;
		self.nonNumber = 5;
		self.showBackground = YES;
		
		srandom(0x7FFFFFFF & time(NULL));
		self.nonsenseController = [[NonsenseSaverController alloc] init];
		self.nonsenses = [NSMutableArray array];
		
		NSFont *theFont = [NSFont systemFontOfSize:kPreviewSize];
		
		for (int i = 0; i < [self nonNumber] ; i++ ) {
			NonsenseObject *non = [[NonsenseObject alloc] initWithString:[self.nonsenseController radomSaying] bounds:[self bounds] font:theFont];
			[self.nonsenses addObject:non];
		}
		
    }
    return self;
}

- (void)reloadScreen:(NSTimer*)thetime
{
	[self setNeedsDisplay:YES];
}

- (void)awakeFromNib
{
	self.refreshRate = [NSTimer timerWithTimeInterval:self.nonDuration target:self selector:@selector(reloadScreen:) userInfo:nil repeats:YES];
}

- (void)drawRect:(NSRect)dirtyRect
{
	[super drawRect:dirtyRect];
	NSFont *theFont = [NSFont systemFontOfSize:kPreviewSize];
	if (self.settingsChanged) {
		self.settingsChanged = NO;
	} else {
		[self.nonsenses removeObjectAtIndex:0];
		NonsenseObject *non = [[NonsenseObject alloc] initWithString:[self.nonsenseController radomSaying] bounds:[self bounds] font:theFont];
		[self.nonsenses addObject:non];
	}
	for (NonsenseObject *obj in self.nonsenses) {
		[obj drawWithBackground:[self showBackground]];
	}
}

@end

File: NONSVisualTester/NonsenseView.swift
//
//  NonsenseView.swift
//  NonsenseSaver
//
//  Created by C.W. Betts on 11/2/14.
//
//

import Cocoa

class NonsenseView: NSView {
	let nonsenseController = NonsenseSaverController()
	dynamic var nonDuration: CGFloat = 2.7 {
		didSet {
			refreshRate?.invalidate()
			refreshRate = Timer(timeInterval: TimeInterval(nonDuration), target: self, selector: #selector(NonsenseView.reloadScreen(_:)), userInfo:nil, repeats: true)
			RunLoop.main.add(refreshRate!, forMode: RunLoopMode.commonModes)
		}
	}
	dynamic var nonNumber: Int = 5 {
		didSet {
			if nonNumber > oldValue {
				
			} else if nonNumber < oldValue {
				
			} else {
				return
			}
			needsDisplay = true
		}
	}
	dynamic var showBackground: Bool = true {
		didSet {
			if oldValue != showBackground {
				needsDisplay = true
			}
		}
	}
	private var nonsenses = [NonsenseObject]()
	private var settingsChanged = false
	private var refreshRate: Timer! = nil

	override func draw(_ dirtyRect: NSRect) {
		super.draw(dirtyRect)
		let theFont = NSFont.systemFont(ofSize: kPreviewFontSize)
		if settingsChanged {
			settingsChanged = false
		} else {
			nonsenses.removeFirst()
			let non = NonsenseObject(string: nonsenseController.randomSaying(), bounds: self.bounds, font: theFont)
			nonsenses.append(non)
		}
		for obj in nonsenses {
			obj.draw(background: showBackground)
		}
	}
	
	func populateNonsenses() {
		let theFont = NSFont.systemFont(ofSize: kPreviewFontSize)
		for _ in 0 ..< nonNumber {
			let non = NonsenseObject(string: nonsenseController.randomSaying(), bounds: self.bounds, font: theFont)
			nonsenses.append(non)
		}
	}
	
	override init(frame: NSRect) {
		srandom(UInt32(time(nil) & 0x7FFFFFFF))
		
		super.init(frame: frame)
		
		populateNonsenses()
	}
	
	required init?(coder: NSCoder) {
		srandom(UInt32(time(nil) & 0x7FFFFFFF))
		
		super.init(coder: coder)
		
		populateNonsenses()
	}
	
	@objc(reloadScreen:) private func reloadScreen(_ theTime: Timer?) {
		self.needsDisplay = true
	}
	
	override func awakeFromNib() {
		super.awakeFromNib()
		
		refreshRate = Timer(timeInterval: TimeInterval(nonDuration), target: self, selector: #selector(NonsenseView.reloadScreen(_:)), userInfo:nil, repeats: true)
		RunLoop.main.add(refreshRate!, forMode: RunLoopMode.commonModes)
	}
	
}
